### DISCUSSION TITLE
Rich Text Editor

### DISCUSSION THREAD

**Discussion by GarrisonJ** [digitallyinduced/ihp#1178] [General]
I'm working on implementing a rich text editor. Quilljs and Trix seem to be the most popular. But, I need to figure out what to save to the DB and how to render it afterward. 

I think the correct way would be to implement a Haskell library to convert the QuillJS-parchment format to HTML, but that isn't very easy. But there are Python/PHP examples. 

Any advice is appreciated.

**Comment by neongreen** [comment]
I have integrated ProseMirror (well, the `tiptap` build) into my app (https://windofchange.me). There's a writeup here: https://windofchange.me/ShowCard?cardId=1d4e2e31-f319-4d01-9798-f775e321fdb1

**Reply by neongreen** [reply]
Right now it works by converting HTML to Markdown on the frontend and then sending that Markdown over to the backend. I might as well send HTML directly — I simply didn't do it because I didn't need it right now.

Untrusted HTML would have to be sanitized (on the backend, not the frontend!) but there are many Haskell libraries available for that. Well, a few. There's definitely at least one: https://hackage.haskell.org/package/xss-sanitize

With Quill, you can just use `innerHTML`: see https://github.com/quilljs/quill/issues/903. You will get HTML with Quill-specific CSS styles apparently, but I personally don't think it's an issue. We use CKEditor at [Brick](https://brick.do) and it follows the same model. But hey, you can always strip out the classes on the frontend/backend if you want. If you do it on the backend, you can use [tagsoup](https://hackage.haskell.org/package/tagsoup).

**Reply by GarrisonJ** [reply]
Nice, thank you.

I actually like the idea of saving it as Markdown because it seems a lot safer. 

I didn't like the idea of using `innerHTML` on the front-end because that would allow the user to send any HTML they wanted.

**Reply by neongreen** [reply]
It doesn't matter. You always have to sanitize things.

Even if the user sends Markdown, there's nothing preventing me from sending

```
foo

<script>...</script>
```

as Markdown — since Markdown supports embedded HTML. Your Markdown renderer might be stripping it out, but you will have to check it.

**Reply by neongreen** [reply]
If you're going to use the `cmark` package, you can sanitize by using `optSafe`:

```haskell
import qualified CMark -- using instead of commonmark because the latter does not support 'safe'

renderMarkdown :: Text -> Html
renderMarkdown text = preEscapedToHtml (CMark.commonmarkToHtml [CMark.optSafe] text)
```

**Reply by neongreen** [reply]
But yeah. If I were you I'd go with Markdown and `optSafe` and only switch to HTML when I actually had to.

**Reply by neongreen** [reply]
Also note that if you do it the way I did it (by copying into a hidden form field), you should pay attention to https://github.com/digitallyinduced/ihp/issues/1177 because otherwise your form submission code might fail and user content would not be saved.

**Reply by neongreen** [reply]
Ping me if you want to integrate TipTap and I'll give you readonly access to my private repo. I don't know how up-to-date my writeup is.

**Reply by GarrisonJ** [reply]
Thanks, I think I'm going to go with that.

**Reply by GarrisonJ** [reply]
I want the toolbar experience, so I think I can use the public library.

**Reply by neongreen** [reply]
Actually I just pasted everything into a gist: https://gist.github.com/neongreen/7dbdddae3af0c476340e0dc175552fad

**Reply by neongreen** [reply]
I don't know if the public library provides a toolbar, btw. Tiptap hates you and wants you to code up the UI by yourself.

There are other ProseMirror builds that have UI — [Remirror](https://remirror.io/) might be the most popular. My approach should work with Remirror as well.

**Comment by nezbite** [comment]
In an internal project we are currently working with TinyMCE for Rich Text editing. It saves to HTML; It's HTML parser sanitizes input when displaying, so that should be safe. This does not prevent users from adding script tags, but anytime I need to display the content, I do it via a TinyMCE editor. If you only want to show the content you can use a read-only TinyMCE editor. In these cases the output is safe. 
The quick and dirty solution (if you don't worry about sanitizing) would be to use TinyMCE for editing and just using `preEscapedToHtml` to render the HTML.
